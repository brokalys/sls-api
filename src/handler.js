import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
  ApolloServerPluginUsageReporting,
  ApolloServerPluginUsageReportingDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';
const isTestMode = process.env.NODE_ENV === 'test';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (isDevMode && !isTestMode) ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (!isTestMode) {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    !isDevMode && !isTestMode
      ? ApolloServerPluginUsageReporting({
          sendVariableValues: {
            exceptNames: ['email', 'unsubscribe_key'],
          },
          logger: {
            debug: () => {},
            info: () => {},
            warn: console.warn,
            error: console.error,
          },
          generateClientInfo: ({ context, request }) => {
            const headers = request.http && request.http.headers;
            const clientVersion =
              headers['apollographql-client-version'] || '1.0.0';

            if (!context.user) {
              return {
                clientName: 'Unauthorized',
                clientVersion,
              };
            }

            return {
              clientName: context.user.apiKey.name,
              clientVersion,
            };
          },
        })
      : ApolloServerPluginUsageReportingDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
