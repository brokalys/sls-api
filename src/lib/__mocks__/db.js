export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
