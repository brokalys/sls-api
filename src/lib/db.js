import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
