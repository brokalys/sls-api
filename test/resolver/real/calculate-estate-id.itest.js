import { createTestClient } from 'apollo-server-testing';
import { server } from 'handler';
import db from 'db-config';

const mockInput = {
  foreign_id: 'id999',
  source: 'ss.lv',
  type: 'sell',
  category: 'apartment',
  url: 'https://example.com',
  price: 10000,
  lat: 56.992294,
  lng: 24.136619,
  cadastre_number: '01000280009',
  images: [],
};

const { mutate } = createTestClient(server);

const MUTATION = `
mutation ChromeExtension_CalculateEstateId(
    $source: String
    $url: String
    $category: String
    $type: String
    $rent_type: String
    $lat: Float
    $lng: Float
    $price: Float
    $price_per_sqm: Float
    $location_district: String
    $location_parish: String
    $location_address: String
    $location_village: String
    $rooms: Int
    $area: Int
    $area_measurement: String
    $floor: Int
    $max_floors: Int
    $content: String
    $building_project: String
    $building_material: String
    $images: [String]!
    $foreign_id: String
    $additional_data: String
    $cadastre_number: String
    $land_area: Float
    $land_area_measurement: String
    $published_at: String
    $views: Int
  ) {
    calculateEstateId(
      source: $source
      url: $url

      category: $category
      type: $type
      rent_type: $rent_type

      lat: $lat
      lng: $lng

      price: $price
      price_per_sqm: $price_per_sqm

      location_district: $location_district
      location_parish: $location_parish
      location_address: $location_address
      location_village: $location_village

      rooms: $rooms
      area: $area
      area_measurement: $area_measurement
      floor: $floor
      max_floors: $max_floors

      content: $content
      building_project: $building_project
      building_material: $building_material
      images: $images

      foreign_id: $foreign_id
      additional_data: $additional_data
      cadastre_number: $cadastre_number

      land_area: $land_area
      land_area_measurement: $land_area_measurement
      published_at: $published_at
      views: $views
    ) {
      id
      type
    }
  }
`;

describe('Mutation: calculateEstateId', () => {
  beforeAll(async () => {
    await db.migrate.latest();
    await db.seed.run();
  });

  test('successfully calculates the estate-id from foreign_id', async () => {
    const response = await mutate({
      mutation: MUTATION,
      variables: {
        ...mockInput,
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('returns null if could not calculate for the property', async () => {
    const response = await mutate({
      mutation: MUTATION,
      variables: {
        ...mockInput,
        foreign_id: null,
      },
    });

    expect(response).toMatchSnapshot();
  });
});
