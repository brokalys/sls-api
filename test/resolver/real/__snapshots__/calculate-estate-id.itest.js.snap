// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation: calculateEstateId returns null if could not calculate for the property 1`] = `
Object {
  "data": Object {
    "calculateEstateId": Object {
      "id": null,
      "type": "building",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: calculateEstateId successfully calculates the estate-id from foreign_id 1`] = `
Object {
  "data": Object {
    "calculateEstateId": Object {
      "id": 4,
      "type": "building",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
